### PR TITLE
SR-10  Add leader endpoints

### PR DESCRIPTION
**Pull request by DocentSzachista**

### CODE REVIEW

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.controller
+
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.services.LeaderService
+
+@RestController
+@RequestMapping("/leaders")
+class Leaders(private val leaderService: LeaderService) {
```
Comment (Mario-659) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt:10]: - change name of the class to LeaderController,
- add `/api/v1` to RequestMapping
- maybe add logger even if not used
- sort and remove unused imports (everywhere)
- add new line at end of file (everywhere)

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/services/LeaderService.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.services
+
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Service
+import org.springframework.web.server.ResponseStatusException
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.model.Leader
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.repository.LeaderRepository
+import java.lang.Exception
+
```
Comment (Mario-659) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/services/LeaderService.kt:9]: - it should be in package service not services

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/repository/LeaderRepository.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.repository
+
+import org.springframework.data.repository.CrudRepository
+import org.springframework.stereotype.Repository
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.model.Leader
+
+@Repository
+interface LeaderRepository : CrudRepository<Leader, Int> {
+
+    fun findByIdAndLedTeamId(leaderId: Int, ledTeamId: Int): Leader?
```
Comment (Mario-659) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/repository/LeaderRepository.kt:10]: typo in findByIdAnd**Led**TeamId

**Review by Mario-659 — CHANGES_REQUESTED** [review]
Fix it 🥺

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/services/LeaderService.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+class LeaderService(var db: LeaderRepository) {
+
+    fun findLeaderInTeam(teamId: Int, leaderId: Int): Leader = db.findByIdAndLedTeamId(leaderId, teamId)
+            ?: throw NoSuchElementException("Leader not found")
+
+    fun findLeadersInTeam(teamId: Int): List<Leader> =
+            db.findByLedTeamId(teamId).ifEmpty {
+                throw NoSuchElementException("Leaders not found")
+            }
+
```
Comment (Mario-659) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/services/LeaderService.kt:21]: add ids to exception messages

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.service
+
+import org.springframework.stereotype.Service
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.model.Leader
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.repository.LeaderRepository
+
+
+@Service
+class LeaderService(var db: LeaderRepository) {
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt:9]: change to val, change name - it's a repository not a database

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+    fun findLeaderInTeam(teamId: Int, leaderId: Int): Leader = db.findByIdAndLeadTeamId(leaderId, teamId)
+            ?: throw NoSuchElementException("Leader $teamId not found")
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt:12]: check if team exists and throw NoSuchElementException with a different message

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    fun findLeadersInTeam(teamId: Int): List<Leader> =
+            db.findByLeadTeamId(teamId).ifEmpty {
+                throw NoSuchElementException("Leaders not found in team $teamId")
+            }
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt:17]: check if team exists and throw NoSuchElementException with a different message

File: backend/recruitment-system/src/main/resources/application.properties [hunk 8]
```diff
@@ -15,4 +15,4 @@ server.error.include-message=always
 
 #Uncomment for better debug
 #spring.jpa.show-sql=true
-#spring.jpa.properties.hibernate.format_sql=true
+#spring.jpa.properties.hibernate.format_sql=true
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/resources/application.properties:18]: Add newline

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.service.LeaderService
+
+@RestController
+@RequestMapping("/api/v1/leaders")
+class Leaders(private val leaderService: LeaderService) {
+
+    private val logger = LoggerFactory.getLogger(this.javaClass)
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt:13]: Add newline after logger

**Review by Mendrejk — CHANGES_REQUESTED** [review]
Fix It!

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt (hunk 1 above)
Comment (Mendrejk) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt:11]: @DocentSzachista you did not change the class name

Comment (DocentSzachista) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leaders.kt]: Done

File: backend/recruitment-system/src/main/resources/application.properties (hunk 8 above)
Comment (Mario-659) [reply backend/recruitment-system/src/main/resources/application.properties:18]: You didn't add newline

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                ?: throw NoSuchElementException("Leader $leaderId not found ")
+    }
+
+    fun findLeadersInTeam(teamId: Int): List<Leader> {
+        assertTeamExists(teamId)
+
+        return leaderRepository.findByLedTeamId(teamId).ifEmpty {
+            throw NoSuchElementException("Leaders not found in team $teamId")
+        }
+
```
Comment (Mario-659) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/LeaderService.kt:25]: I can see you are throwing an exception when list is empty. Are you sure about that? Shouldn't we just return empty list?

**Review by Mario-659 — CHANGES_REQUESTED** [review]

**Review by Mario-659 — DISMISSED** [review]

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leader.kt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.service.LeaderService
+
+@RestController
+@RequestMapping("/api/v1/leaders")
+class Leader(private val leaderService: LeaderService) {
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leader.kt:11]: The class (and the file) should be called LeaderController

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leader.kt [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+    private val logger = LoggerFactory.getLogger(this.javaClass)
+
+    @ExceptionHandler(NoSuchElementException::class)
+    fun handleNotFound(e: NoSuchElementException): ResponseEntity<String> = ResponseEntity(e.message, HttpStatus.NOT_FOUND)
+
+    @GetMapping("/{team_id}/{leader_id}")
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leader.kt:18]: The full endpoint should be
/api/v1/**teams**/{team_id}/**leaders**/{leader_id}
Yours is
/api/v1/leaders/{team_id}/{leader_id}

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leader.kt [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    fun leader(@PathVariable team_id: Int, @PathVariable leader_id: Int) = leaderService.findLeaderInTeam(team_id, leader_id)
+
+    @GetMapping("/{team_id}")
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/Leader.kt:21]: The full endpoint should be
/api/v1/**teams**/{team_id}/**leaders**
Yours is
/api/v1/leaders/{team_id}

**Review by Mendrejk — CHANGES_REQUESTED** [review]
Fix It!

**Review by Mendrejk — APPROVED** [review]
Ship It!

**Review by Mario-659 — APPROVED** [review]

**Review by deamacto — APPROVED** [review]

**Pull request merged**